Operator crashes when storage backend cannot be detected

Seems #58 is back, seeing this on BYOR control plane cluster, no cloudprovider set:
```
I0111 01:01:08.021751       1 main.go:24] Cluster Image Registry Operator Version: 2e125da-dirty
I0111 01:01:08.021867       1 main.go:25] Go Version: go1.10.3
I0111 01:01:08.021872       1 main.go:26] Go OS/Arch: linux/amd64
I0111 01:01:08.049742       1 controller.go:378] waiting for informer caches to sync
I0111 01:01:09.251837       1 controller.go:387] started events processor
I0111 01:01:09.263690       1 bootstrap.go:102] generating registry custom resource
E0111 01:01:09.267640       1 storage.go:106] unknown storage backend: 
E0111 01:01:09.267778       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/bootstrap.go:128
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:122
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:213
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:220
/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:385
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x12918ad]

goroutine 96 [running]:
github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x144be20, 0x223d8d0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).Bootstrap(0xc42002c160, 0xc420810480, 0x1)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/bootstrap.go:128 +0x6ad
github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).sync(0xc42002c160, 0x17f2570, 0xc4200c7460)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:122 +0x1e6
github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).eventProcessor.func1(0xc42002c160, 0x13b60a0, 0x17bb930)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:213 +0x8f
github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).eventProcessor(0xc42002c160)
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:220 +0x8e
github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).(github.com/openshift/cluster-image-registry-operator/pkg/operator.eventProcessor)-fm()
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:385 +0x2a
github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc420aa80a0)
	/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc420aa80a0, 0x3b9aca00, 0x0, 0x9c295639fd35a501, 0xc4200b0de0)
	/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc420aa80a0, 0x3b9aca00, 0xc4200b0de0)
	/go/src/github.com/openshift/cluster-image-registry-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/openshift/cluster-image-registry-operator/pkg/operator.(*Controller).Run
	/go/src/github.com/openshift/cluster-image-registry-operator/pkg/operator/controller.go:385 +0xc32
```

--- a/pkg/operator/bootstrap.go
+++ b/pkg/operator/bootstrap.go
@@ -125,7 +125,10 @@ func (c *Controller) Bootstrap() error {
 	}
 
 	modified := false
-	err = driver.CompleteConfiguration(cr, &modified)
+	err = nil
+	if driver != nil {
+		err = driver.CompleteConfiguration(cr, &modified)
+	}
 	_, cerr := client.Configs().Create(cr)
 	if cerr != nil {
 		return cerr

--- a/pkg/resource/generator.go
+++ b/pkg/resource/generator.go
@@ -104,9 +104,7 @@ func (g *Generator) syncStorage(cr *imageregistryv1.Config, modified *bool) erro
 		return err
 	}
 
-	if len(driver.GetStorageName()) == 0 {
-		runCreate = true
-	} else if driver.StorageChanged(cr, modified) {
+	if driver.StorageChanged(cr, modified) {
 		runCreate = true
 	} else {
 		exists, err := driver.StorageExists(cr, modified)

--- a/pkg/storage/emptydir/emptydir.go
+++ b/pkg/storage/emptydir/emptydir.go
@@ -86,6 +86,7 @@ func (d *driver) GetStorageName() string {
 func (d *driver) CreateStorage(cr *imageregistryv1.Config, modified *bool) error {
 	if !reflect.DeepEqual(cr.Status.Storage.Filesystem, cr.Spec.Storage.Filesystem) {
 		cr.Status.Storage.Filesystem = d.Config.DeepCopy()
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionTrue, "Creation Successful", "EmptyDir storage successfully created", modified)
 		*modified = true
 	}
 

--- a/pkg/storage/s3/s3.go
+++ b/pkg/storage/s3/s3.go
@@ -437,10 +437,26 @@ func (d *driver) RemoveStorage(cr *imageregistryv1.Config, modified *bool) error
 		return err
 	}
 
-	cr.Spec.Storage.S3.Bucket = ""
+	// Wait until the bucket does not exist
+	if err := svc.WaitUntilBucketNotExists(&s3.HeadBucketInput{
+		Bucket: aws.String(d.Config.Bucket),
+	}); err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionTrue, aerr.Code(), aerr.Error(), modified)
+		}
+
+		return err
+	}
+
+	if len(cr.Spec.Storage.S3.Bucket) != 0 {
+		cr.Spec.Storage.S3.Bucket = ""
+		*modified = true
+	}
+
 	d.Config.Bucket = ""
-	cr.Status.Storage.S3 = d.Config.DeepCopy()
-	if d.StorageChanged(cr, modified) {
+
+	if !reflect.DeepEqual(cr.Status.Storage.S3, d.Config) {
+		cr.Status.Storage.S3 = d.Config.DeepCopy()
 		*modified = true
 	}
 

--- a/pkg/testframework/framework.go
+++ b/pkg/testframework/framework.go
@@ -16,7 +16,7 @@ var (
 	// operations to complete. ForeverTestTimeout is not long enough to create
 	// several replicas and get them available on a slow machine.
 	// Setting this to 5 minutes:w
-	
+
 	AsyncOperationTimeout = 5 * time.Minute
 )
 

